File: chapter 09/recipe 9.7/DnScTranScope/DniScTranScopeObj.cs
using System;
using System.Transactions;

using TransactionLogging;

namespace DniScTranScope
{
	public interface ILocalScope
	{
		void UseLocalScope(bool succeed);
	}

	public class DniScTranScopeObj : ILocalScope
	{
		public void UseLocalScope(bool succeed)
		{
			TransactionScope scope = new TransactionScope();
			using (scope)
			{
				//log the state of the transaction
				TransactionLogger log
					= new TransactionLogger(this);

				//
				//update a database or other 
				//resource manager here
				//

				if (succeed)
				{
					//signal that the transaction 
					//should be committed
					scope.Complete();
				}
			}
		}
	}
}

File: chapter 09/recipe 9.7/DniScTranScopeVB/DniScTranScopeObj.vb
Imports System.Transactions
Imports DniScTranScopeVB.TransactionLogging

Public Interface ILocalScope
    Sub UseLocalScope(ByVal succeed As Boolean)
End Interface

Public Class DniScTranScopeObj
    Implements ILocalScope

    Public Sub UseLocalScope(ByVal succeed As Boolean) _
            Implements ILocalScope.UseLocalScope

        'start the transaction scope
        Using scope As TransactionScope _
            = New TransactionScope()

            'log the transaction
            Dim log As TransactionLogger _
                = New TransactionLogger(Me)

            '
            'update a database or other 
            'resource manager here
            '

            If succeed Then
                scope.Complete()
            End If
        End Using

    End Sub
End Class
